src/thread.rs

```diff
@@ -0,0 +1,483 @@
+        None
+    }
+
+    /// Change the active thread to some enabled thread.
+    fn yield_active_thread(&mut self) {
+        self.yield_active_thread = true;
+    }
+
+    /// Decide which action to take next and on which thread.
+    fn schedule(&mut self) -> InterpResult<'tcx, SchedulingAction> {
+        if self.threads[self.active_thread].check_terminated() {
+            // Check if we need to unblock any threads.
+            for (i, thread) in self.threads.iter_enumerated_mut() {
+                if thread.state == ThreadState::BlockedOnJoin(self.active_thread) {
+                    trace!("unblocking {:?} because {:?} terminated", i, self.active_thread);
+                    thread.state = ThreadState::Enabled;
+                }
+            }
+            return Ok(SchedulingAction::ExecuteDtors);
+        }
+        if self.threads[self.active_thread].state == ThreadState::Enabled
+            && !self.yield_active_thread
+        {
+            return Ok(SchedulingAction::ExecuteStep);
+        }
+        for (id, thread) in self.threads.iter_enumerated() {
+            if thread.state == ThreadState::Enabled {
+                if !(self.yield_active_thread && id == self.active_thread) {
+                    self.active_thread = id;
+                    break;
+                }
+            }
+        }
+        self.yield_active_thread = false;
+        if self.threads[self.active_thread].state == ThreadState::Enabled {
+            return Ok(SchedulingAction::ExecuteStep);
+        }
+        if self.threads.iter().all(|thread| thread.state == ThreadState::Terminated) {
+            Ok(SchedulingAction::Stop)
+        } else {
+            throw_machine_stop!(TerminationInfo::Deadlock);
+        }
+    }
+}
+
+impl<'mir, 'tcx: 'mir> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
+    /// A workaround for thread-local statics until
+    /// https://github.com/rust-lang/rust/issues/70685 is fixed: change the
+    /// thread-local allocation id with a freshly generated allocation id for
+    /// the currently active thread.
+    fn remap_thread_local_alloc_ids(
+        &self,
+        val: &mut mir::interpret::ConstValue<'tcx>,
+    ) -> InterpResult<'tcx> {
+        let this = self.eval_context_ref();
+        match val {
+            mir::interpret::ConstValue::Scalar(Scalar::Ptr(ptr)) => {
+                let alloc_id = ptr.alloc_id;
+                let alloc = this.tcx.alloc_map.lock().get(alloc_id);
+                let tcx = this.tcx;
+                let is_thread_local = |def_id| {
+                    tcx.codegen_fn_attrs(def_id).flags.contains(CodegenFnAttrFlags::THREAD_LOCAL)
+                };
+                match alloc {
+                    Some(GlobalAlloc::Static(def_id)) if is_thread_local(def_id) => {
+                        ptr.alloc_id = this.get_or_create_thread_local_alloc_id(def_id)?;
```

Comment:
Wow, this is really subtle... `ptr` is actually a reference due to match ergonomics, and here we implicitly dereference it. I'd prefer if we could make this a little more explicit. Any suggestions?

Reply:
```rust
        match *val {
            mir::interpret::ConstValue::Scalar(Scalar::Ptr(ref mut ptr)) => {
```

Reply:
I have adapted the @bjorn3 suggestion because I could not think of anything better.